Mathlib/NumberTheory/NumberField/Completion.lean

```diff
@@ -0,0 +1,339 @@
+automatically infer these. A common approach to handle the ambiguity that arises from having
+multiple sources of instances is through the use of type synonyms. In this case, we define a
+type synonym `WithAbs` for a semiring which depends on an absolute value. This provides a
+systematic way of assigning and inferring instances of the semiring that depend on an absolute
+value. In our application, relevant instances and the completion of a number field `K` are first
+defined at the level of `AbsoluteValue` by using the type synonym `WithAbs` of `K`, and then
+derived downstream for `InfinitePlace` (which is a subtype of `AbsoluteValue`).
+
+## Main definitions
+ - `WithAbs` : type synonym equipping a semiring with an absolute value.
+ - `AbsoluteValue.completion` : the uniform space completion of a field `K` equipped with real
+  absolute value.
+ - `NumberField.InfinitePlace.completion` : the completion of a number field `K` at an infinite
+  place, obtained by completing `K` with respect to the absolute value associated to the infinite
+  place.
+ - `NumberField.InfinitePlace.Completion.extensionEmbedding` : the embedding `v.embedding : K →+* ℂ`
+  extended to `v.completion →+* ℂ`.
+ - `NumberField.InfinitePlace.Completion.extensionEmbedding_of_isReal` : if the infinite place `v`
+  is real, then this extends the embedding `v.embedding_of_isReal : K →+* ℝ` to
+  `v.completion →+* ℝ`.
+ - `NumberField.InfinitePlace.Completion.equiv_real_of_isReal` : the ring isomorphism
+  `v.completion ≃+* ℝ` when `v` is a real infinite place; the forward direction of this is
+  `extensionEmbedding_of_isReal`.
+ - `NumberField.InfinitePlace.Completion.equiv_complex_of_isComplex` : the ring isomorphism
+  `v.completion ≃+* ℂ` when `v` is a complex infinite place; the forward direction of this is
+  `extensionEmbedding`.
+
+## Main results
+ - `NumberField.Completion.locallyCompactSpace` : the completion of a number field at
+  an infinite place is locally compact.
+ - `NumberField.Completion.isometry_extensionEmbedding` : the embedding `v.completion →+* ℂ` is
+  an isometry. See also `isometry_extensionEmbedding_of_isReal` for the corresponding result on
+  `v.completion →+* ℝ` when `v` is real.
+ - `NumberField.Completion.bijective_extensionEmbedding_of_isComplex` : the embedding
+  `v.completion →+* ℂ` is bijective when `v` is complex. See also
+  `bijective_extensionEmebdding_of_isReal` for the corresponding result for `v.completion →+* ℝ`
+  when `v` is real.
+
+## Tags
+number field, embeddings, infinite places, completion
+-/
+noncomputable section
+
+/-- Type synonym equipping a semiring with an absolute value. -/
+@[nolint unusedArguments]
+def WithAbs {R S : Type*} [Semiring R] [OrderedSemiring S] :
+    AbsoluteValue R S → Type _ := fun _ => R
+
+namespace WithAbs
+
+variable {K : Type*} [Field K] (v : AbsoluteValue K ℝ)
+
+instance normedField : NormedField (WithAbs v) :=
```

Comment:
If you make it for a `NormedRing`, is there a chance that in the case of fields everything goes smoothly or do you need to re-do everything? In the first case, it is worth generalising it, but in the second case of course no, leave it as it is.

Reply:
This is possible, but it will require a bit of extra work. Further down in the file I want to show that the completion of a number field at an infinite place is indeed a field, so I'm using [this](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/UniformField.html#UniformSpace.Completion.instField), which requires a `TopologicalDivisionRing` instance on `K`. This comes for free if `K` is a `NormedField` but not if `K` is a `NormedRing`. On the other hand, for the non-Archimedean case, `Valued` class carries a [`TopologicalDivisionRing`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Valued/ValuedField.html#Valued.topologicalDivisionRing) instance so it all works smoothly using `Valued` there.